libs/@local/hash-graph-types/typescript/src/entity.ts

```diff
@@ -83,22 +83,56 @@ export type PropertyMetadataElement =
   | PropertyMetadataObject
   | PropertyMetadataValue;
 
+export const isValueMetadata = (
+  metadata: PropertyMetadataElement,
+): metadata is PropertyMetadataValue => !("value" in metadata);
+
+export const isArrayMetadata = (
+  metadata: PropertyMetadataElement,
+): metadata is PropertyMetadataArray =>
+  !isValueMetadata(metadata) && Array.isArray(metadata.value);
+
+export const isObjectMetadata = (
+  metadata: PropertyMetadataElement,
+): metadata is PropertyMetadataObject =>
+  !isValueMetadata(metadata) && !Array.isArray(metadata.value);
+
 export interface PropertyMetadataValue {
   metadata: Omit<ValueMetadata, "dataTypeId"> & {
     dataTypeId?: VersionedUrl;
   };
 }
 
 export interface PropertyMetadataObject {
-  value?: Record<BaseUrl, PropertyMetadataElement>;
+  value: Record<BaseUrl, PropertyMetadataElement>;
```

Comment:
I made `value` required to distinguish between the variants. `PropertyMetadataValue` does not have `value`.

Reply:
I think the fact that we use `value` and `metadata` at different levels for different things might get confusing and I generally like to avoid `something.metadata.metadata`
```ts
const propertyObject: PropertyObjectWithMetadata = {
  value: { } // the property object
  metadata: {
    value: {} // the metadata for the object values
    metadata: {} // the metadata for the object itself
  }
}

const array: PropertyArrayWithMetadata = {
  value: [] // the array
  metadata: {
    value: [] // the metadata for the array elements
    metadata:  // the metadata for the array itself
  }
}
```
Maybe we can rename them?
```ts
export interface PropertyMetadataArray {
  values: PropertyMetadataElement[];
  array?: ArrayMetadata;
}

export interface PropertyMetadataObject {
  values: Record<BaseUrl, PropertyMetadataElement>;
  object?: ObjectMetadata;
}
```

Reply:
We don't have any nested metadata, e.g.:
```typescript
{
  value: {
    "https://.../top-level-property-1/": { // Top level property with leaf value
      value: "foo", // Leaf value, only present if the properties are embedded into the object
      metadata: { // metadata of top-level-property-1
        dataTypeId: ...
        provenance: ...
        ...
      }
    },
    "https://.../top-level-property-2/": { // Top level property with nested properties
      value: {
        "https://.../nested-in-property-2/": { // Nested property
          value: "bar",
          metadata: { // metadata of nested-in-property-2
            dataTypeId: ...
            provenance: ...
            ...
          }
        }
      },
      metadata: { // metadata of top-level-property-2
        provenance: ...
        ...
      }
    },
    "https://.../top-level-property-3/": { // Top level property with array of values
      value: [
        { // first element
          value: "element 1",
          metadata: { // metadata of nested-in-property-2
            dataTypeId: ...
            provenance: ...
            ...
          }
        },
        { // second element
          value: "element 2",
          metadata: { // metadata of nested-in-property-2
            dataTypeId: ...
            provenance: ...
            ...
          }
        }
      ],
      metadata: { // metadata of top-level-property-3
        provenance: ...
        ...
      }
    },
  },
  metadata: { // metadata for whole property object
    provenance: ...
    ...
  }
}
```
The difference between `PropertyObjectWithMetadata` and `PropertyMetadataObject` is that the leaf values also contains a `value`.

We, however, could instead use 
- `value` for leaf values as defined by data types
- `array` for array properties
- `object` for nested properties, so the above would become:

```typescript
{
  object: {
    "https://.../top-level-property-1/": { // Top level property with leaf value
      value: "foo", // Leaf value, only present if the properties are embedded into the object
      metadata: { // metadata of top-level-property-1
        dataTypeId: ...
        provenance: ...
        ...
      }
    },
    "https://.../top-level-property-2/": { // Top level property with nested properties
      object: {
        "https://.../nested-in-property-2/": { // Nested property
          value: "bar",
          metadata: { // metadata of nested-in-property-2
            dataTypeId: ...
            provenance: ...
            ...
          }
        }
      },
      metadata: { // metadata of top-level-property-2
        provenance: ...
        ...
      }
    },
    "https://.../top-level-property-3/": { // Top level property with array of values
      array: [
        { // first element
          value: "element 1",
          metadata: { // metadata of nested-in-property-2
            dataTypeId: ...
            provenance: ...
            ...
          }
        },
        { // second element
          value: "element 2",
          metadata: { // metadata of nested-in-property-2
            dataTypeId: ...
            provenance: ...
            ...
          }
        }
      ],
      metadata: { // metadata of top-level-property-3
        provenance: ...
        ...
      }
    },
  },
  metadata: { // metadata for whole property object
    provenance: ...
    ...
  }
}
```

Reply:
What is `PropertyMetadataObject["value"]` being used for?

Reply:
Can we add comments to these types explaining what they are? This would help

Reply:
It's here to distinguish between `metadata` and inner values. It would be possible to flatten the inner properties but that does not work for arrays, so I chose to use `value` for both. Also, it's easier to provide better errors when deserializing malformed objects and forbidding any non-allowed keys (flattening the `value` basically means using `additionalProperties: true`.

Reply:
> Can we add comments to these types explaining what they are? This would help

Yes, absolutely!

Reply:
We spoke about this, the confusion came from there being two representations:
1. When the metadata is defined separately (`properties` object just contains values) – this is `PropertyMetadataObject` / `PropertyMetadataArray` – this representation is only the type for input at the moment, e.g. create and update, although in practice we don't yet actually use it
2. When the metadata appears under `metadata` alongside the value under `value` – this is the other types, as discussed above, and is what we pass to the graph (but is not exposed to any consumer)

Let's leave it for now as it's going to change – we can add comments and clarify as we go